### DISCUSSION TITLE
Getting conflicts while integrating Matplotplusplus in a MFC SDI single Document application architecture project with visual studio 2022

### DISCUSSION THREAD

**Discussion by OTJack** [alandefreitas/matplotplusplus#216] [Q&A]
Hi,

I'm developing a test application for Matploplusplus in C++/Visual Studio 2022 on Windows 10. The application is a MFC SDI single Document application architecture project compiles and links very well alone but when I integrate the library with <matplot/maplot.h> I get tones of conflict compiling errors:
Note that I customized the project setting with the appropriate include and library path, and also I setup the compiler to the latest C++ (20).
I tried two way to get the latest Matplot++ library distributions release, one built with CMake by myself with success and one already built and downloaded @ https://github.com/alandefreitas/matplotplusplus/releases
In both cases, I got the same errors.

1) First type of error were min max macro definitions. The file "Common.h" uses a min/max definitions which conflict with Microsoft min/max macros defined in <minwindef.h>. I googled to find if there were ways to fix the pb and it was suggested to define NOMINMAX before "Windows.h"... but the application doesn't show this include and so I defined this pre processor key word in the project setting. Unfortunately, defining the keyword during the whole compilation process create an error for missing max macro definition....

Here is the error message:
Severity	Code	Description							Project	File										Line	Suppression State
Warning		C4003	not enough arguments for function-like macro invocation 'min'	MFC_TestApp-v3	C:\Program Files\matplotplusplus 1.1.0\include\matplot\util\common.h	101	

Is there another way to circonvent the problem?

2) Second: The second type of error I got is related to the "iterator" definition:

Severity	Code	Description										Project	File										Line	Suppression State
Warning	       	C4099	'std::iterator': type name first seen using 'struct' now seen using 'class'		MFC_TestApp-v3	C:\Program Files\matplotplusplus 1.1.0\include\matplot\util\common.h	543	
Error		C2990	'std::iterator': non-class template has already been declared as a class template	MFC_TestApp-v3	C:\Program Files\matplotplusplus 1.1.0\include\matplot\util\common.h	543	

Iterator appears as a class in the "common.h" file, but apparently it's already defined in C++ as per the litterature:
https://docs.microsoft.com/en-us/cpp/standard-library/iterators?view=msvc-170

Is there any documentation/fix about this conflict yet?

__________________________________________________________
        double operator()(std::size_t i, std::size_t j) {
            return _vec.at(i).at(j);
        }
        friend class iterator;

class iterator {
            vector_2d_view *const _vec;
            std::size_t _row_offset;
            std::size_t _col_offset;

            static bool is_empty_iterator(const iterator &it) {
                return it._vec == nullptr;
__________________________________________________________

Note that when I integrate the Matplotplusplus in a Windows console application (not MFC), the build is successful, with no error.

Thank you for your support
Jacques

**Comment by alandefreitas** [comment]
There are a number of solutions for `NOMINMAX`. You can `undef` min/max yourself before including matplot.h or, if you want to open a PR, you can adjust the code to allow min/max macros and then replace whatever calls of min/max with (min)/(max), which is what boost does. For instance, `std::max(a,b)` would become `(std::max)(a,b)`.

As for the iterator, they shouldn't conflict because they are in different namespaces. If I'm not wrong, C4099 should only be a warning. Assuming you're not `using namespace std;`, this should be OK.
